tokio-util/src/io/sync_read_stream.rs

```diff
@@ -0,0 +1,111 @@
+use futures_core::stream::Stream;
+use std::io;
+use std::io::Read;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::task::Context;
+use std::task::Poll;
+use tokio::runtime::Handle;
+
+const DEFAULT_BUFFER_SIZE: usize = 1024 * 16;
+type EmptyOrBuffer = Option<Vec<u8>>;
+type SyncReadResult = Result<EmptyOrBuffer, io::Error>;
+
+/// Use a [`SyncReadStream`] to asynchronously read from a
+/// synchronous API.
+#[derive(Debug)]
+pub struct SyncReadStream<R: Read + Send> {
+    reader: Arc<Mutex<R>>,
+    waiting_answer: Arc<Mutex<Option<SyncReadResult>>>,
```

Comment:
Why not use a `tokio::sync::mpsc::Receiver<bytes::Bytes>` here?
It would be much easier to implement `SyncReadStream` itself while also reusing the memory if possible.

Reply:
Good point, I was not sure I could pull it using a receiver, but you are totally right.

Reply:
[tokio::sync::mpsc::Receiver::poll_recv](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.poll_recv) can be used to poll the receiver.

There's also block_recv for using it in blocking thread.

Reply:
I pushed an alternative version, I'm not sure it improved drastically the code but got rid of the clone of the waker and other stuff. Though, I am not sure `bytes::Bytes` is that useful here.